network/p2p/tracer/gossipSubMeshTracer.go

```diff
@@ -0,0 +1,184 @@
+func (t *GossipSubMeshTracer) Graft(p peer.ID, topic string) {
+	t.topicMeshMu.Lock()
+	defer t.topicMeshMu.Unlock()
+
+	lg := t.logger.With().Str("topic", topic).Str("peer_id", p.String()).Logger()
+
+	if _, ok := t.topicMeshMap[topic]; !ok {
+		t.topicMeshMap[topic] = make(map[peer.ID]struct{})
+	}
+	t.topicMeshMap[topic][p] = struct{}{}
+
+	id, exists := t.idProvider.ByPeerID(p)
+	if !exists {
+		lg.Warn().
+			Bool(logging.KeySuspicious, true).
+			Msg("grafted peer not found in identity provider")
+		return
+	}
+
+	lg.Info().Hex("flow_id", logging.ID(id.NodeID)).Str("role", id.Role.String()).Msg("grafted peer")
+	t.metrics.OnLocalMeshSizeUpdated(topic, len(t.topicMeshMap[topic]))
+}
+
+// Prune is called when a peer is removed from a topic mesh. The tracer uses this to track the mesh peers.
+func (t *GossipSubMeshTracer) Prune(p peer.ID, topic string) {
+	t.topicMeshMu.Lock()
+	defer t.topicMeshMu.Unlock()
+
+	lg := t.logger.With().Str("topic", topic).Str("peer_id", p.String()).Logger()
+
+	if _, ok := t.topicMeshMap[topic]; !ok {
+		return
+	}
+	delete(t.topicMeshMap[topic], p)
+
+	id, exists := t.idProvider.ByPeerID(p)
+	if !exists {
+		lg.Warn().
+			Bool(logging.KeySuspicious, true).
+			Msg("pruned peer not found in identity provider")
+		return
+	}
+
+	lg.Info().Hex("flow_id", logging.ID(id.NodeID)).Str("role", id.Role.String()).Msg("pruned peer")
+	t.metrics.OnLocalMeshSizeUpdated(topic, len(t.topicMeshMap[topic]))
+}
+
+// logLoop logs the mesh peers of the local node for each topic at a regular interval.
+func (t *GossipSubMeshTracer) logLoop(ctx irrecoverable.SignalerContext) {
+	ticker := time.NewTicker(t.loggerInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			t.logPeers()
+		}
+	}
+}
+
+// logPeers logs the mesh peers of the local node for each topic.
+// Note that based on GossipSub parameters, we expect to have between 6 and 12 peers in the mesh for each topic.
+// Hence, choosing a heartbeat interval in the order of minutes should be sufficient to log the mesh peers of the local node.
+// Also, note that the mesh peers are also logged reactively when a peer is added or removed from the mesh.
+func (t *GossipSubMeshTracer) logPeers() {
+	t.topicMeshMu.RLock()
+	defer t.topicMeshMu.RUnlock()
+
+	for topic := range t.topicMeshMap {
+		shouldWarn := false // whether we should warn about the mesh state
+		lg := t.logger.With().Dur("heartbeat_interval", t.loggerInterval).Str("topic", topic).Logger()
+		for p := range t.topicMeshMap[topic] {
+			id, exists := t.idProvider.ByPeerID(p)
+			if !exists {
+				shouldWarn = true
+				lg = lg.With().
+					Str("peer_id", p.String()).
+					Str("flow_id", "unknown").
+					Str("role", "unknown").
+					Logger()
+
+				continue
+			}
+
+			lg = lg.With().
+				Str("peer_id", p.String()).
```

Comment:
This appears to be adding `peer_id`, `flow_id`, and `role` for each of the 6-12 topic peers. a couple of questions:
1. What does zerolog do when it gets duplicate keys? append/overwrite/something else? would it make sense to use a [dictionary](https://github.com/rs/zerolog#sub-dictionary)?
2. how many characters will these log lines be? most systems have a max after which they are truncated.

Reply:
This code logs the local peers on each topic mesh separately, and prints a log entry per topic. The `zerolog` dictionary seems applicable to cases where we dump the entire topic mesh topology in a single log, which helps group the log entries based on the keys and merge the duplicates. I think having a separate log entry per topic (as the current approach) provides more clarity and simplifies debugging that dumping all topology-related info in the same log. 

Also, looking at a single log, the total number of characters for the keys and values in the key-value pair breakdown are:
`peer_id`: key has 7 characters, value has 52 characters
`flow_id`: key has 7 characters, value has 64 characters
`role`: key has 4 characters, value has 9 characters
Therefore, in the topic-based approach, adding each peer entry takes the total number of 143 characters, i.e., the total number of characters for the keys is 18 (7 + 7 + 4) and the total number of characters for the values is 125 (52 + 64 + 9).
Each log entry in this way also has around 70 characters for general meta-data and messages including the log level. So, with `x` peers in a topic mesh, the topic-based logging takes `70 + 125 * x`. This is around `720` characters for `6`-nodes topic meshes and around `1570` characters for `12-node` topic meshes. Accordingly, I think the current approach is safe with respect to truncation.

Reply:
> There will be no duplicate key on a log

I mean that there will be up to 12 `peer_id` entries in the log line right? I _think_ zerolog will allow this, and that it will serialize all labels in the order they were set, but i'm not sure. Have you tested it? Using a dictionary would make sure the groupings are preserved even if the log consumer doesn't maintain the original order.